### PR TITLE
Deepro dev

### PR DESCRIPTION
**Pull request by TerminalVelocityDPro**
Wrote eval_net_motion.py (which is a edited copy of eval_net.py) to run inference on test set scans with 2D motion corruption via the methodology outlined in previous Google Colabs. To do this, motion sweep values were removed from eval_net_motion.py and build_recon_val_loader was modified to induce 2D Motion Corruption. This was done by modifying the build_recon_val_loader function in meddlr/data/build.py by adding angle, translation, nshots, and trajectory parameters and by replacing DataTransform with a newly made MotionDataTransform that takes in the angle, translation, nshots, and trajectory values specified in the parameters. MotionDataTransform is written in meddlr/data/transforms/transform.py and copies much of the original DataTransform. This biggest and most important change is the replacement of MotionModel with MotionModel2D, which takes in nshots, angle, translation, and trajectory values specified. MotionModel2D conducts the 2D Motion Corruption and is written in meddlr/data/transforms/2D_motion.py. It copies largely from MotionModel but simulates 2D motion for multi-shot Cartesian MRI instead. 

Notes:
This is completely untested and undebugged code. 

Please let me know what you think / what I need to fix / what I could improve. In addition, let me know if there are any places that are confusing and that need better commenting, or if there are areas that need better documentation. 

Thanks,
Deepro.

### CODE REVIEW

File: meddlr/data/build.py [hunk 1]
```diff
@@ -243,13 +243,16 @@ def build_recon_train_loader(cfg, dataset_type=None):
...
 def build_recon_val_loader(
     cfg,
     dataset_name,
     as_test: bool = False,
     add_noise: bool = False,
     add_motion: bool = False,
+    angle: float = 0,
+    translation: float = 0,
+    nshots: int = 0,
+    trajectory: str = "blocked",
```
Comment (ad12) [inline meddlr/data/build.py]: we want to avoid taking in too many arguments - consider passing in a `data_transform`

File: meddlr/data/transforms/2D_motion.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+class MotionModel2D:
+    """A model that corrupts kspace inputs with motion.
+
+    Motion is a common artifact experienced during the MR imaging forward problem.
+    When a patient moves, the recorded (expected) location of the kspace sample is
+    different than the actual location where the kspace sample that was acquired.
+    This module is responsible for simulating different motion artifacts.
+    """
+
+    def __init__(self, nshots, angle, translate, trajectory, seed=None):
```
Comment (ad12) [inline meddlr/data/transforms/2D_motion.py]: Add type annotations. For example `nshots: int`

File: meddlr/data/transforms/2D_motion.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        """
+        Args:
+            nshots (int) : The number of shots in the image. 
+                This should be equivalent to ceil(phase_encode_dim / 
+                echo_train_length). 
+            angle : The (min, max) angle for rotation. Values should be in 
+                degrees and should be >=-180, <=180. Use 'None' to
```
Comment (ad12) [inline meddlr/data/transforms/2D_motion.py]: nit: use range notation instead `[-180, 180]`

Comment (ad12) [reply meddlr/data/transforms/2D_motion.py]: if this can be `None`, make sure it is listed as an optional variable: `angle: Optional[Tuple[int, int]]`. Given that it is optional, it should default to `None`

File: meddlr/data/transforms/2D_motion.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            nshots (int) : The number of shots in the image. 
+                This should be equivalent to ceil(phase_encode_dim / 
+                echo_train_length). 
+            angle : The (min, max) angle for rotation. Values should be in 
+                degrees and should be >=-180, <=180. Use 'None' to 
+                ignore rotation.
+            translate: The fraction of (height, width) to translate. 
+                e.g. 0.1 => 10% of the corresponding dimension.
+                So (0.1, 0.2) => 10% of height, 20% of width.
+                Use 'None' to ignore translation.
```
Comment (ad12) [inline meddlr/data/transforms/2D_motion.py]: See comment for `angle`

File: meddlr/data/transforms/2D_motion.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+        This function supports two trajectories:
+        - 'blocked' : Where each shot corresponds to a consecutive block of 
+          kspace. (e.g. 1 1 2 2 3 3)
+        - 'interleaved' : Where shots are interleaved (e.g. 1 2 3 1 2 3) 
+        
+        We assume the phase encode direction is left to right (i.e. along
+        width dimesion). 
+
+        TODO: Add support for sensitivity maps.
```
Comment (ad12) [inline meddlr/data/transforms/2D_motion.py]: these comments don't apply for this function - move them to the class docstring level

File: meddlr/data/transforms/2D_motion.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        Args:
+            image : The complex-valued iamge. Shape [..., height, width].
+
+        Returns:
+            The motion corrupted kspace. 
+        """
+        if self.seed == None:
+            random_motion = RandomAffine
+        else:
+            random_motion = RandomAffine.seed(self.seed)
```
Comment (ad12) [inline meddlr/data/transforms/2D_motion.py]: I don't think you can seed the RandomAffine like this - `seed` is an instance method, not a class method

File: meddlr/data/transforms/2D_motion.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import numpy as np
+import torch
+import meddlr.ops as F
+from meddlr.transforms import RandomAffine, RandomTranslation, TransformList
+from meddlr.forward.mri import SenseModel
+
+from meddlr.utils.events import get_event_storage
+
+
+class MotionModel2D:
```
Comment (ad12) [inline meddlr/data/transforms/2D_motion.py]: It seems like this class is just calling the function we wrote - what is the advantage of having a class here? There doesn't seem to be any state we are preserving.

If there is no state, just add the function to your eval script for now and call that in the MotionDataTransform

File: meddlr/data/transforms/transform.py [hunk 8]
```diff
@@ -11,6 +11,7 @@
 from meddlr.utils import transforms as T
 
 from .motion import MotionModel
+from .motion import MotionModel2D
```
Comment (ad12) [inline meddlr/data/transforms/transform.py]: use absolute paths, not relative paths, for imports - eg. `from meddlr.data.transforms.motion import MotionModel2D`

File: meddlr/data/transforms/transform.py [hunk 9]
```diff
@@ -364,3 +365,211 @@ def __call__(self, kspace, maps, target, fname, slice_id, is_fixed, acceleration
...
+    def __init__(
+        self,
+        cfg,
+        mask_func,
+        is_test: bool = False,
+        add_noise: bool = False,
+        add_motion: bool = False,
+        angle: float = 0,
+        translation: float = 0,
+        nshots: int = 0,
```
Comment (ad12) [inline meddlr/data/transforms/transform.py]: is a default value of `0` ever valid? if not, set it to `None` and raise a ValueError if it is not set

File: meddlr/data/transforms/transform.py [hunk 10]
```diff
@@ -364,3 +365,211 @@ def __call__(self, kspace, maps, target, fname, slice_id, is_fixed, acceleration
         target = target.squeeze(0)
 
         return masked_kspace, maps, target, mean, std, norm
+
+
+class MotionDataTransform:
```
Comment (ad12) [inline meddlr/data/transforms/transform.py:372]: We are only going to use this transform for inference right now - only keep the logic that is relevant when `is_test=True`

File: meddlr/data/transforms/transform.py [hunk 11]
```diff
@@ -364,3 +365,211 @@ def __call__(self, kspace, maps, target, fname, slice_id, is_fixed, acceleration
...
+        maps = out["maps"]
+        target = out["target"]
+        mean = out["mean"]
+        std = out["std"]
+
+        # Get rid of batch dimension...
+        masked_kspace = masked_kspace.squeeze(0)
+        maps = maps.squeeze(0)
+        target = target.squeeze(0)
+
```
Comment (ad12) [inline meddlr/data/transforms/transform.py]: i dont think you need this function for your transform class

File: meddlr/data/transforms/transform.py [hunk 12]
```diff
@@ -364,3 +365,211 @@ def __call__(self, kspace, maps, target, fname, slice_id, is_fixed, acceleration
...
+        mean = normalized["mean"]
+        std = normalized["std"]
+
+        add_noise = self.add_noise and (
+            self._is_test or (not is_fixed and self.rng.uniform() < self.p_noise)
+        )
+        add_motion = self.add_motion and (
+            self._is_test or (not is_fixed and self.rng.uniform() < self.p_motion)
+        )
+        if add_motion:
```
Comment (ad12) [inline meddlr/data/transforms/transform.py]: motion simulation should happen before the undersampling

File: tools/eval_net_motion.py [hunk 13]
```diff
@@ -0,0 +1,549 @@
+"""Run inference on test set scans.
```
Comment (ad12) [inline tools/eval_net_motion.py:1]: thinking more about it, it may make sense to build your motion pipeline into the original `eval_net.py` file. it looks like the changes should be relatively simple to that file, because most of the heavy lifting is being done by the transform

**Review by ad12 — CHANGES_REQUESTED** [review]

**Comment by ad12** [comment]
Also make sure to run the linter and formatter - run `make autoformat` from the meddlr repository folder

File: meddlr/data/transforms/2D_motion.py (hunk 7 above)
Comment (TerminalVelocityDPro) [reply meddlr/data/transforms/2D_motion.py]: So, I'm taking the function that we wrote and I'm writing it within the MotionDataTransform. I'm erasing this class as a result. Is that correct or was I supposed to do something else? Thanks!

Comment (TerminalVelocityDPro) [reply meddlr/data/transforms/2D_motion.py]: @ad12

Comment (ad12) [reply meddlr/data/transforms/2D_motion.py]: Correct - remove the class. Note, if the function depends on instance attributes in `MotionDataTransform`, make it an instance method. If it does not, make it a standalone function

File: meddlr/data/transforms/transform.py (hunk 9 above)
Comment (TerminalVelocityDPro) [reply meddlr/data/transforms/transform.py]: @ad12 Should this ValueError be raised if a parameter is set to None? Sorry, just trying to better understand. I added an Optional to the angle and translation parameters for the motion corruption function, so the function can handle None parameters for angle and translation right? So, would I still raise that ValueError? Or do I raise it for something else?

Comment (TerminalVelocityDPro) [reply meddlr/data/transforms/transform.py]: Would nshots be checked to see if it was set? And thus would that raise a ValueError? Also, would the type annotation for nshots in this function be just nshots: int? Thanks, Deepro.

File: meddlr/data/transforms/transform.py (hunk 12 above)
Comment (TerminalVelocityDPro) [reply meddlr/data/transforms/transform.py]: Where should the noise simulation go? Before or after the undersampling? Thanks, Deepro.

File: meddlr/data/transforms/transform.py (hunk 10 above)
Comment (TerminalVelocityDPro) [reply meddlr/data/transforms/transform.py:372]: Should something be done to handle for when is_test is False? Or is that not necessary for now? Thanks.

File: meddlr/data/build.py (hunk 1 above)
Comment (TerminalVelocityDPro) [reply meddlr/data/build.py]: If I pass in just a data_transform as a parameter (for instance like MotionDataTransform), how would I pass in the parameters for angle, translation, nshots, and trajectory into the MotionDataTransform? Thanks, Deepro.

File: meddlr/data/build.py [hunk 14]
```diff
@@ -250,6 +250,12 @@ def build_recon_val_loader(
     as_test: bool = False,
     add_noise: bool = False,
     add_motion: bool = False,
+    two_dimensional: bool = True,
```
Comment (ad12) [inline meddlr/data/build.py]: Take this in as an integer - we may be 4D/5D/6D etc in the future.

```
mri_dim = 2
```

File: meddlr/data/build.py [hunk 15]
```diff
@@ -276,8 +282,17 @@ def build_recon_val_loader(
...
+        cfg,
+        mask_func,
+        is_test=as_test,
+        add_noise=add_noise,
+        add_motion=add_motion,
+        two_dimensional=two_dimensional,
+        angle=angle,
+        translation=translation,
+        nshots=nshots,
+        trajectory=trajectory,
```
Comment (ad12) [inline meddlr/data/build.py]: Given that there are multiple data transforms that we can have, i would recommend passing in the transform as a argument to this method. 

What you have done here is scrapped the old DataTransform for yours, so people will not be able to use the old DataTransform anymore. Instead, have the user be able to optionally pass in a `transform`. If `transform` is not passed in, default to the `DataTransform`

Comment (ad12) [reply meddlr/data/build.py]: So in your eval script, you would pass in a MotionDataTransform

File: meddlr/data/transforms/motion_corruption_2D.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
+import math
+from typing import Optional
+
+import torch
+
+import meddlr.ops as F
+from meddlr.transforms import RandomAffine
+
+
+def add_motion_corruption(
```
Comment (ad12) [inline meddlr/data/transforms/motion_corruption_2D.py]: Move this function to this file

https://github.com/ad12/meddlr/blob/main/meddlr/transforms/functional/mri.py

File: meddlr/data/transforms/transform.py [hunk 17]
```diff
@@ -364,3 +367,183 @@ def __call__(self, kspace, maps, target, fname, slice_id, is_fixed, acceleration
...
+        self.trajectory = trajectory
+
+        seed = cfg.SEED if cfg.SEED > -1 else None
+        self.rng = np.random.RandomState(seed)
+
+        if is_test:
+            # When we test we dont want to initialize with certain parameters (e.g. scheduler).
+            self.noiser = NoiseModel(cfg.MODEL.CONSISTENCY.AUG.NOISE.STD_DEV, seed=seed)
+        else:
+            pass
```
Comment (ad12) [inline meddlr/data/transforms/transform.py:434]: If `'is_test` must be `True`, have a check at the beginning of `__init__` that raises a `ValueError` with a detailed description if that is not the case

File: meddlr/data/transforms/transform.py [hunk 18]
```diff
@@ -364,3 +367,183 @@ def __call__(self, kspace, maps, target, fname, slice_id, is_fixed, acceleration
...
+        norm = torch.sqrt(torch.mean(cplx.abs(target) ** 2))
+
+        # TODO: Add other transforms here.
+
+        # If 2D MRI, then each slice will have different motion - seed is some
+        # combination of the file name and the slice id (+).
+        # If 3D MRI, then each slice should have the same motion - seed is 
+        # determined only by the file name.  
+        if self.two_dimensional:
+            seed = sum(tuple(map(ord, fname))) + slice if self._is_test else None  # noqa
```
Comment (ad12) [inline meddlr/data/transforms/transform.py]: what is `slice`? should this be `slice_id`?

File: meddlr/data/transforms/transform.py [hunk 19]
```diff
@@ -364,3 +367,183 @@ def __call__(self, kspace, maps, target, fname, slice_id, is_fixed, acceleration
...
+
+        add_motion = self.add_motion and self._is_test
+        if add_motion:
+            # Motion seed should not be different for each slice for now.
+            # TODO: Change this for 2D acquisitions.
+
+            tfm_gen = RandomAffine(p=1.0, translate=self.translation, angle=self.angle)
+            tfm_gen.seed(seed)
+
+            kspace = add_motion_corruption(
```
Comment (ad12) [inline meddlr/data/transforms/transform.py:511]: Because the scan is undersampled, we want to make sure we are not adding motion to the entries in the k-space that are zero. Remember the scan is undersampled, so there will be zero-value entries in the kspace. 

A hacky way to do this for now is to perturb the scan with motion and then mask it out with the undersampling mask again so that zero-valued entries stay zero-valued. 

A more realistic way to do this is to handle it in the `add_motion_corruption` function. In this function, you would only consider lines that have some non-zero value. So if you took the sum along the height dimension of the mask, the valid lines would be those that are non-zero

File: tools/eval_net.py [hunk 20]
```diff
@@ -458,6 +470,46 @@ def main(args):
...
+        ),
+    )
+
+    parser.add_argument(
+        "--two-dimensional",
+        default="true",
+        choice=("true", "false"),
+        help=("If true, MRI is 2D. Otherwise, MRI is 3D.")
+    )
+
```
Comment (ad12) [inline tools/eval_net.py]: this is actually dataset specific - fastMRI is 2D, mridata is 3D. Note your motion simulation does not support 3D yet, so i would actually just raise an error if it is not the fastMRI dataset you are using. or at least make sure to only use the fastmri dataset

**Review by ad12 — CHANGES_REQUESTED** [review]

File: meddlr/data/transforms/transform.py (hunk 18 above)
Comment (TerminalVelocityDPro) [reply meddlr/data/transforms/transform.py]: yep it should be slice_id, thank you!
